File: template.py
class Product:
    def __init__(self, name, id, price):
        self._name = name
        self._id = id
        self._price = price

    def get_name(self):
        return self._name

    def set_name(self, name):
        self._name = name

    def get_id(self):
        return self._id

    def set_id(self, id):
        self._id = id

    def get_price(self):
        return self._price

    def set_price(self, price):
        self._price = price

    def visualizeAtHome(self, photo):
        pass


class ClothesProducts(Product):
    def __init__(self, name, id, price, isNew):
        super().__init__(name, id, price)
        self._isNew = isNew

    def get_isNew(self):
        return self._isNew

    def set_isNew(self, isNew):
        self._isNew = isNew


class FurnitureProducts(Product):
    def __init__(self, name, id, price, height, weight, width):
        super().__init__(name, id, price)
        self._height = height
        self._weight = weight
        self._width = width

    def get_height(self):
        return self._height

    def set_height(self, height):
        self._height = height

    def get_weight(self):
        return self._weight

    def set_weight(self, weight):
        self._weight = weight

    def get_width(self):
        return self._width

    def set_width(self, width):
        self._width = width


class VegetablesProducts(Product):
    def __init__(self, name, id, price, freshness):
        super().__init__(name, id, price)
        self._freshness = freshness

    def get_freshness(self):
        return self._freshness

    def set_freshness(self, freshness):
        self._freshness = freshness


class MilkProducts(Product):
    def __init__(self, name, id, price):
        super().__init__(name, id, price)


products = {
    FurnitureProducts("3 Legs Chair", 12345, 100,10,101,20),
    FurnitureProducts("Sofa - Black", 12346, 200,10,101,20),
    FurnitureProducts("Sofa - Red ", 12347, 250,10,101,20),
    FurnitureProducts("Table - Vintage", 12348, 100,10,101,20),
    ClothesProducts("T-shirt", 12349, 200,True)
}

for product in products:
    print("The name of the product is now", product.get_name(), ",price is:", product.get_price())

File: template.java
# create discount system
# no need to serialze to/from DB - 
# never change for each product seperatly - create rules
# first roll 10% discount for all products
# change the structure just once (not in secod phase) 
# no cart 

public class Product {

    private String name;
    private int id;
    private double price;

    public Product(String name, int id, double price) {
        this.name = name;
        this.id = id;
        this.price = price;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public double getPrice() {
        return price;
    }

    public void setPrice(double price) {
        this.price = price;
    }

    // No implementation needed for visualizeAtHome in Java

}

class ClothesProduct extends Product {

    private boolean isNew;

    public ClothesProduct(String name, int id, double price, boolean isNew) {
        super(name, id, price);
        this.isNew = isNew;
    }

    public boolean getIsNew() {
        return isNew;
    }

    public void setIsNew(boolean isNew) {
        this.isNew = isNew;
    }
}

class FurnitureProduct extends Product {

    private double height;
    private double weight;
    private double width;

    public FurnitureProduct(String name, int id, double price, double height, double weight, double width) {
        super(name, id, price);
        this.height = height;
        this.weight = weight;
        this.width = width;
    }

    public double getHeight() {
        return height;
    }

    public void setHeight(double height) {
        this.height = height;
    }

    public double getWeight() {
        return weight;
    }

    public void setWeight(double weight) {
        this.weight = weight;
    }

    public double getWidth() {
        return width;
    }

    public void setWidth(double width) {
        this.width = width;
    }
}

class VegetablesProduct extends Product {

    private double freshness;

    public VegetablesProduct(String name, int id, double price, double freshness) {
        super(name, id, price);
        this.freshness = freshness;
    }

    public double getFreshness() {
        return freshness;
    }

    public void setFreshness(double freshness) {
        this.freshness = freshness;
    }
}

class MilkProduct extends Product {

    public MilkProduct(String name, int id, double price) {
        super(name, id, price);
    }
}

public class DiscountSystem {

    public static void main(String[] args) {
        Map<Integer, Product> products = new HashMap<>();
        products.put(12345, new FurnitureProduct("3 Legs Chair", 12345, 100, 10, 101, 20));
        products.put(12346, new FurnitureProduct("Sofa - Black", 12346, 200, 10, 101, 20));
        products.put(12347, new FurnitureProduct("Sofa - Red ", 12347, 250, 10, 101, 20));
        products.put(12348, new FurnitureProduct("Table - Vintage", 12348, 100, 10, 101, 20));
        products.put(12349, new ClothesProduct("T-shirt", 12349, 200, true));

        for (Product product : products.values()) {
            System.out.println("The name of the product is now " + product.getName() + ", price is: " + product.getPrice());
            // Implement discount logic here (e.g., product.setPrice(product.getPrice() * 0.9); for 10% discount)
        }
    }
}
